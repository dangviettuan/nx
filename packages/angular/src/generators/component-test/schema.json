{
  "$schema": "http://json-schema.org/schema",
  "$id": "NxAngularComponentTestGenerator",
  "type": "object",
  "cli": "nx",
  "description": "Create a `*.cy.ts` file for Cypress component testing for an Angular component.",
  "properties": {
    "project": {
      "type": "string",
      "description": "The name of the project where the component is located.",
      "x-dropdown": "projects",
      "x-prompt": "What project is the component located in?"
    },
    "componentName": {
      "type": "string",
      "description": "Class name of the component to create a test for.",
      "examples": ["MyFancyButtonComponent"]
    },
    "componentDir": {
      "type": "string",
      "description": "Relative path to the folder that contains the component from the project root.",
      "example": ["src/lib/my-fancy-button"]
    },
    "componentFileName": {
      "type": "string",
      "description": "File name that contains the component without the `.ts` extension.",
      "examples": ["my-fancy-button.component"]
    },
    "skipFormat": {
      "description": "Skip formatting files.",
      "type": "boolean",
      "default": false
    }
  },
  "additionalProperties": false,
  "required": [
    "projectPath",
    "componentName",
    "componentPath",
    "componentFileName"
  ]
}
