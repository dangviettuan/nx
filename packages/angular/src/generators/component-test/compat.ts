import { convertNxGenerator } from '@nrwl/devkit';
import { componentTestGenerator } from './component-test';

export default convertNxGenerator(componentTestGenerator);
