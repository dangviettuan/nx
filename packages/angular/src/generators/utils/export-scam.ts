import type { Tree } from '@nrwl/devkit';
import {
  joinPathFragments,
  logger,
  names,
  readProjectConfiguration,
  stripIndents,
} from '@nrwl/devkit';
import { locateLibraryEntryPointFromDirectory } from './entry-point';
import type { FileInfo } from './file-info';
import { getRelativeImportToFile } from './path';

export type GenerationOptions = {
  name: string;
  project: string;
  export?: boolean;
  inlineScam?: boolean;
};

export function exportScam(
  tree: Tree,
  fileInfo: FileInfo,
  options: GenerationOptions
): void {
  if (!options.export) {
    return;
  }

  const { projectType, root, sourceRoot } = readProjectConfiguration(
    tree,
    options.project
  );
  const projectSourceRoot = sourceRoot ?? joinPathFragments(root, 'src');

  if (projectType === 'application') {
    logger.warn(
      '--export=true was ignored as the project the SCAM is being generated in is not a library.'
    );

    return;
  }

  const entryPointPath = locateLibraryEntryPointFromDirectory(
    tree,
    fileInfo.directory,
    root,
    projectSourceRoot
  );
  if (!entryPointPath) {
    // Let's not error, simply warn the user
    // It's not too much effort to manually do this
    // It would be more frustrating to have to find the correct path and re-run the command
    logger.warn(
      `Could not export SCAM. Unable to determine project's entry point. SCAM has still been created.`
    );

    return;
  }

  const relativePathFromEntryPoint = getRelativeImportToFile(
    entryPointPath,
    fileInfo.filePath
  );
  const entryPointContent = tree.read(entryPointPath, 'utf-8');
  let updatedEntryPointContent = stripIndents`${entryPointContent}
    export * from "${relativePathFromEntryPoint}";`;

  if (!options.inlineScam) {
    const moduleFilePath = joinPathFragments(
      fileInfo.directory,
      `${names(options.name).fileName}.module.ts`
    );
    const relativePathFromModule = getRelativeImportToFile(
      entryPointPath,
      moduleFilePath
    );
    updatedEntryPointContent = stripIndents`${updatedEntryPointContent}
        export * from "${relativePathFromModule}";`;
  }

  tree.write(entryPointPath, updatedEntryPointContent);
}
