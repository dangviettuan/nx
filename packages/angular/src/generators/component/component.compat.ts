import componentGenerator from './component';
import { warnForSchematicUsage } from '../utils/warn-for-schematic-usage';
import { convertNxGenerator } from '@nrwl/devkit';

export default warnForSchematicUsage(convertNxGenerator(componentGenerator));
