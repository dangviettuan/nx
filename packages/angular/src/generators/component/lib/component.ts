import type { Tree } from '@nrwl/devkit';
import { logger, readProjectConfiguration, stripIndents } from '@nrwl/devkit';
import { getComponentFileInfo } from '../../utils/file-info';
import { locateLibraryEntryPointFromDirectory } from '../../utils/entry-point';
import { getRelativeImportToFile } from '../../utils/path';
import type { NormalizedSchema } from '../schema';
import { shouldExportInEntryPoint } from './entry-point';
import { findModuleFromOptions } from './module';

export function exportComponentInEntryPoint(
  tree: Tree,
  schema: NormalizedSchema
): void {
  if (!schema.export || (schema.skipImport && !schema.standalone)) {
    return;
  }

  const { root, projectType } = readProjectConfiguration(tree, schema.project);

  if (projectType === 'application') {
    return;
  }

  const { directory, filePath } = getComponentFileInfo(tree, schema);

  const entryPointPath = locateLibraryEntryPointFromDirectory(
    tree,
    directory,
    root,
    schema.projectSourceRoot
  );
  if (!entryPointPath) {
    logger.warn(
      `Unable to determine whether the component should be exported in the library entry point file. ` +
        `The library's entry point file could not be found. Skipping exporting the component in the entry point file.`
    );

    return;
  }

  if (!schema.standalone) {
    const modulePath = findModuleFromOptions(tree, schema, root);
    if (!shouldExportInEntryPoint(tree, entryPointPath, modulePath)) {
      return;
    }
  }

  const relativePathFromEntryPoint = getRelativeImportToFile(
    entryPointPath,
    filePath
  );
  const updateEntryPointContent = stripIndents`${tree.read(
    entryPointPath,
    'utf-8'
  )}
    export * from "${relativePathFromEntryPoint}";`;

  tree.write(entryPointPath, updateEntryPointContent);
}
