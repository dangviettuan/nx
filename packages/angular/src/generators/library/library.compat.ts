import { convertNxGenerator } from '@nrwl/devkit';
import { warnForSchematicUsage } from '../utils/warn-for-schematic-usage';
import library from './library';

export const librarySchematic = warnForSchematicUsage(
  convertNxGenerator(library)
);
