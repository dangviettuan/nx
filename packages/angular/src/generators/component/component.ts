import type { Tree } from '@nrwl/devkit';
import {
  formatFiles,
  normalizePath,
  readProjectConfiguration,
  readWorkspaceConfiguration,
} from '@nrwl/devkit';
import { wrapAngularDevkitSchematic } from '@nrwl/devkit/ngcli-adapter';
import { pathStartsWith } from '../utils/path';
import { exportComponentInEntryPoint } from './lib/component';
import { normalizeOptions } from './lib/normalize-options';
import type { NormalizedSchema, Schema } from './schema';

export async function componentGenerator(tree: Tree, rawOptions: Schema) {
  const options = normalizeOptions(tree, rawOptions);
  const { projectSourceRoot, ...schematicOptions } = options;

  checkPathUnderProjectRoot(tree, options);

  const angularComponentSchematic = wrapAngularDevkitSchematic(
    '@schematics/angular',
    'component'
  );
  await angularComponentSchematic(tree, schematicOptions);

  exportComponentInEntryPoint(tree, options);

  await formatFiles(tree);
}

function checkPathUnderProjectRoot(tree: Tree, schema: NormalizedSchema): void {
  if (!schema.path) {
    return;
  }

  const project =
    schema.project ?? readWorkspaceConfiguration(tree).defaultProject;
  const { root } = readProjectConfiguration(tree, project);

  let pathToComponent = normalizePath(schema.path);
  pathToComponent = pathToComponent.startsWith('/')
    ? pathToComponent.slice(1)
    : pathToComponent;

  if (!pathStartsWith(pathToComponent, root)) {
    throw new Error(
      `The path provided for the component (${schema.path}) does not exist under the project root (${root}). ` +
        `Please make sure to provide a path that exists under the project root.`
    );
  }
}

export default componentGenerator;
