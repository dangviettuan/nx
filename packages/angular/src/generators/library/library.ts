import {
  addDependenciesToPackageJson,
  formatFiles,
  installPackagesTask,
  moveFilesToNewDirectory,
  removeDependenciesFromPackageJson,
  Tree,
} from '@nrwl/devkit';
import { wrapAngularDevkitSchematic } from '@nrwl/devkit/ngcli-adapter';
import { jestProjectGenerator } from '@nrwl/jest';
import { Linter } from '@nrwl/linter';
import { convertToNxProjectGenerator } from '@nrwl/workspace/generators';
import init from '../../generators/init/init';
import { E2eTestRunner } from '../../utils/test-runners';
import { ngPackagrVersion } from '../../utils/versions';
import addLintingGenerator from '../add-linting/add-linting';
import karmaProjectGenerator from '../karma-project/karma-project';
import setupTailwindGenerator from '../setup-tailwind/setup-tailwind';
import { addBuildableLibrariesPostCssDependencies } from '../utils/dependencies';
import { addModule } from './lib/add-module';
import {
  enableStrictTypeChecking,
  setLibraryStrictDefault,
} from './lib/enable-strict-type-checking';
import { normalizeOptions } from './lib/normalize-options';
import { NormalizedSchema } from './lib/normalized-schema';
import { updateLibPackageNpmScope } from './lib/update-lib-package-npm-scope';
import { updateProject } from './lib/update-project';
import { updateTsConfig } from './lib/update-tsconfig';
import { addStandaloneComponent } from './lib/add-standalone-component';
import { Schema } from './schema';

export async function libraryGenerator(tree: Tree, schema: Partial<Schema>) {
  // Do some validation checks
  if (!schema.routing && schema.lazy) {
    throw new Error(`To use "--lazy" option, "--routing" must also be set.`);
  }

  if (schema.publishable === true && !schema.importPath) {
    throw new Error(
      `For publishable libs you have to provide a proper "--importPath" which needs to be a valid npm package name (e.g. my-awesome-lib or @myorg/my-lib)`
    );
  }

  if (schema.addTailwind && !schema.buildable && !schema.publishable) {
    throw new Error(
      `To use "--addTailwind" option, you have to set either "--buildable" or "--publishable".`
    );
  }

  const options = normalizeOptions(tree, schema);
  const { libraryOptions, componentOptions } = options;

  await init(tree, {
    ...libraryOptions,
    skipFormat: true,
    e2eTestRunner: E2eTestRunner.None,
  });

  const runAngularLibrarySchematic = wrapAngularDevkitSchematic(
    '@schematics/angular',
    'library'
  );
  await runAngularLibrarySchematic(tree, {
    name: libraryOptions.name,
    prefix: libraryOptions.prefix,
    entryFile: 'index',
    skipPackageJson:
      libraryOptions.skipPackageJson ||
      !(libraryOptions.publishable || libraryOptions.buildable),
    skipTsConfig: true,
  });

  if (libraryOptions.ngCliSchematicLibRoot !== libraryOptions.projectRoot) {
    moveFilesToNewDirectory(
      tree,
      libraryOptions.ngCliSchematicLibRoot,
      libraryOptions.projectRoot
    );
  }
  await updateProject(tree, libraryOptions);
  updateTsConfig(tree, libraryOptions);
  await addUnitTestRunner(tree, libraryOptions);
  updateNpmScopeIfBuildableOrPublishable(tree, libraryOptions);

  if (!libraryOptions.standalone) {
    addModule(tree, libraryOptions);
  } else {
    await addStandaloneComponent(tree, options);
  }

  setStrictMode(tree, libraryOptions);
  await addLinting(tree, libraryOptions);

  if (libraryOptions.addTailwind) {
    await setupTailwindGenerator(tree, {
      project: libraryOptions.name,
      skipFormat: true,
    });
  }

  if (libraryOptions.buildable || libraryOptions.publishable) {
    removeDependenciesFromPackageJson(tree, [], ['ng-packagr']);
    addDependenciesToPackageJson(
      tree,
      {},
      {
        'ng-packagr': ngPackagrVersion,
      }
    );
    addBuildableLibrariesPostCssDependencies(tree);
  }

  if (libraryOptions.standaloneConfig) {
    await convertToNxProjectGenerator(tree, {
      project: libraryOptions.name,
      all: false,
      skipFormat: true,
    });
  }

  if (!libraryOptions.skipFormat) {
    await formatFiles(tree);
  }

  return () => {
    installPackagesTask(tree);
  };
}

async function addUnitTestRunner(
  host: Tree,
  options: NormalizedSchema['libraryOptions']
) {
  if (options.unitTestRunner === 'jest') {
    await jestProjectGenerator(host, {
      project: options.name,
      setupFile: 'angular',
      supportTsx: false,
      skipSerializers: false,
      skipFormat: true,
    });
  } else if (options.unitTestRunner === 'karma') {
    await karmaProjectGenerator(host, {
      project: options.name,
      skipFormat: true,
    });
  }
}

function updateNpmScopeIfBuildableOrPublishable(
  host: Tree,
  options: NormalizedSchema['libraryOptions']
) {
  if (options.buildable || options.publishable) {
    updateLibPackageNpmScope(host, options);
  }
}

function setStrictMode(
  host: Tree,
  options: NormalizedSchema['libraryOptions']
) {
  if (options.strict) {
    enableStrictTypeChecking(host, options);
  } else {
    setLibraryStrictDefault(host, options.strict);
  }
}

async function addLinting(
  host: Tree,
  options: NormalizedSchema['libraryOptions']
) {
  if (options.linter === Linter.None) {
    return;
  }
  await addLintingGenerator(host, {
    projectName: options.name,
    projectRoot: options.projectRoot,
    prefix: options.prefix,
    unitTestRunner: options.unitTestRunner,
    setParserOptionsProject: options.setParserOptionsProject,
    skipFormat: true,
  });
}

export default libraryGenerator;
