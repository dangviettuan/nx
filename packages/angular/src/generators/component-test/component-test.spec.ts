import { assertMinimumCypressVersion } from '@nrwl/cypress/src/utils/cypress-version';
import { Tree } from '@nrwl/devkit';
import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';
import { Linter } from '@nrwl/linter';
import { UnitTestRunner } from '../../utils/test-runners';
import { componentGenerator } from '../component/component';
import { libraryGenerator } from '../library/library';
import { componentTestGenerator } from './component-test';
jest.mock('@nrwl/cypress/src/utils/cypress-version');
describe('Angular Cypress Component Test Generator', () => {
  let tree: Tree;
  let mockedAssertMinimumCypressVersion: jest.Mock<
    ReturnType<typeof assertMinimumCypressVersion>
  > = assertMinimumCypressVersion as never;
  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace();
    // silence warnings about missing .gitignore file
    tree.write('.gitignore', '');
    mockedAssertMinimumCypressVersion.mockReturnValue();
  });

  it('should handle component w/o inputs', async () => {
    await libraryGenerator(tree, {
      name: 'my-lib',
      unitTestRunner: UnitTestRunner.None,
      linter: Linter.None,
    });
    await componentGenerator(tree, {
      project: 'my-lib',
      name: 'my-lib',
    });
    componentTestGenerator(tree, {
      componentName: 'MyLibComponent',
      componentFileName: './my-lib.component',
      project: 'my-lib',
      componentDir: 'src/lib/my-lib',
    });
    expect(
      tree.read('libs/my-lib/src/lib/my-lib/my-lib.component.cy.ts', 'utf-8')
    ).toMatchSnapshot();
  });

  it('should generate a component test', async () => {
    await libraryGenerator(tree, {
      name: 'my-lib',
      unitTestRunner: UnitTestRunner.None,
      linter: Linter.None,
    });
    await componentGenerator(tree, {
      project: 'my-lib',
      name: 'my-lib',
    });

    tree.write(
      'libs/my-lib/src/lib/my-lib/my-lib.component.ts',
      `
import { Component, OnInit, Input } from '@angular/core';

export type ButtonStyle = 'default' | 'primary' | 'accent';

@Component({
  selector: 'proj-my-lib',
  templateUrl: './my-lib.component.html',
  styleUrls: ['./my-lib.component.css']
})
export class MyLibComponent implements OnInit {
  @Input('buttonType') type = 'button';
  @Input() style: ButtonStyle = 'default';
  @Input() age?: number;
  @Input() isOn = false;
  @Input() message: string | undefined;
  @Input() anotherProp: any;
  @Input() anotherNeverProp: never;
  
  constructor() { }

  ngOnInit(): void {
  }

}`
    );

    componentTestGenerator(tree, {
      componentName: 'MyLibComponent',
      componentFileName: './my-lib.component',
      project: 'my-lib',
      componentDir: 'src/lib/my-lib',
    });

    expect(
      tree.read('libs/my-lib/src/lib/my-lib/my-lib.component.cy.ts', 'utf-8')
    ).toMatchSnapshot();
  });

  it('should work with standalone components', async () => {
    await libraryGenerator(tree, {
      name: 'my-lib',
      unitTestRunner: UnitTestRunner.None,
      linter: Linter.None,
    });
    await componentGenerator(tree, {
      project: 'my-lib',
      name: 'my-lib',
      standalone: true,
    });
    tree.write(
      'libs/my-lib/src/lib/my-lib/my-lib.component.ts',
      `
import { Component, OnInit, Input } from '@angular/core';
import { CommonModule } from '@angular/common';

@Component({
  selector: 'proj-my-lib',
  standalone: true,
  imports: [CommonModule],
  templateUrl: './my-lib.component.html',
  styleUrls: ['./my-lib.component.css']
})
export class MyLibComponent implements OnInit {
  @Input('buttonType') type = 'button';
  @Input() style: ButtonStyle = 'default';
  @Input() age?: number;
  @Input() isOn = false;
  @Input() message: string | undefined;
  @Input() anotherProp: any;
  @Input() anotherNeverProp: never;
  constructor() { }

  ngOnInit(): void {
  }

}
`
    );
    componentTestGenerator(tree, {
      componentName: 'MyLibComponent',
      componentFileName: './my-lib.component',
      project: 'my-lib',
      componentDir: 'src/lib/my-lib',
    });
    expect(
      tree.read('libs/my-lib/src/lib/my-lib/my-lib.component.cy.ts', 'utf-8')
    ).toMatchSnapshot();
  });

  it('should not overwrite an existing component test', async () => {
    await libraryGenerator(tree, {
      name: 'my-lib',
      unitTestRunner: UnitTestRunner.None,
      linter: Linter.None,
    });

    await componentGenerator(tree, { name: 'my-lib', project: 'my-lib' });
    tree.write(
      'libs/my-lib/src/lib/my-lib/my-lib.component.cy.ts',
      `should not overwrite`
    );

    componentTestGenerator(tree, {
      componentName: 'MyLibComponent',
      componentFileName: './my-lib.component',
      project: 'my-lib',
      componentDir: 'src/lib/my-lib',
    });

    expect(
      tree.read('libs/my-lib/src/lib/my-lib/my-lib.component.cy.ts', 'utf-8')
    ).toEqual('should not overwrite');
  });

  it('should be idempotent', async () => {
    await libraryGenerator(tree, {
      name: 'my-lib',
      unitTestRunner: UnitTestRunner.None,
      linter: Linter.None,
    });

    await componentGenerator(tree, { name: 'my-lib', project: 'my-lib' });

    const expected = `import { MountConfig, mount } from 'cypress/angular';
import { MyLibComponent } from './my-lib.component';

describe(MyLibComponent.name, () => {
  const config: MountConfig<MyLibComponent> = {
    declarations: [],
    imports: [],
    providers: []
  }

  it('renders', () => {
     mount(MyLibComponent, config);
  })
})
`;

    componentTestGenerator(tree, {
      componentName: 'MyLibComponent',
      componentFileName: './my-lib.component',
      project: 'my-lib',
      componentDir: 'src/lib/my-lib',
    });
    expect(
      tree.read('libs/my-lib/src/lib/my-lib/my-lib.component.cy.ts', 'utf-8')
    ).toEqual(expected);

    componentTestGenerator(tree, {
      componentName: 'MyLibComponent',
      componentFileName: './my-lib.component',
      project: 'my-lib',
      componentDir: 'src/lib/my-lib',
    });
    expect(
      tree.read('libs/my-lib/src/lib/my-lib/my-lib.component.cy.ts', 'utf-8')
    ).toEqual(expected);
  });
});
