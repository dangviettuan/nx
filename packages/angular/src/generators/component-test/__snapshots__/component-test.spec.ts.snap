// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Angular Cypress Component Test Generator should generate a component test 1`] = `
"import { MountConfig, mount } from 'cypress/angular';
import { MyLibComponent } from './my-lib.component';

describe(MyLibComponent.name, () => {
  const config: MountConfig<MyLibComponent> = {
    declarations: [],
    imports: [],
    providers: []
  }

  it('renders', () => {
     mount(MyLibComponent, {
           ...config,
           componentProperties: {
               type:  'button',
               style:  'default',
               age:  0,
               isOn:  false,
          }
       });
  })
})
"
`;

exports[`Angular Cypress Component Test Generator should handle component w/o inputs 1`] = `
"import { MountConfig, mount } from 'cypress/angular';
import { MyLibComponent } from './my-lib.component';

describe(MyLibComponent.name, () => {
  const config: MountConfig<MyLibComponent> = {
    declarations: [],
    imports: [],
    providers: []
  }

  it('renders', () => {
     mount(MyLibComponent, config);
  })
})
"
`;

exports[`Angular Cypress Component Test Generator should work with standalone components 1`] = `
"import { MountConfig, mount } from 'cypress/angular';
import { MyLibComponent } from './my-lib.component';

describe(MyLibComponent.name, () => {
  const config: MountConfig<MyLibComponent> = {
    declarations: [],
    imports: [],
    providers: []
  }

  it('renders', () => {
     mount(MyLibComponent, {
           ...config,
           componentProperties: {
               type:  'button',
               style:  'default',
               age:  0,
               isOn:  false,
          }
       });
  })
})
"
`;
